workitem/workitem_repository.go

```diff
@@ -701,50 +702,182 @@ func (r *GormWorkItemRepository) Fetch(ctx context.Context, spaceID uuid.UUID, c
-// 		count( case fields->>'system.state' when 'closed' then '1' else null end ) as Closed
-// 		FROM "work_items" left join iterations
-// 		on fields@> concat('{"system.iteration": "', iterations.id, '"}')::jsonb
-// 		WHERE (iterations.space_id = '33406de1-25f1-4969-bcec-88f29d0a7de3'
-// 		and work_items.deleted_at IS NULL) GROUP BY IterationId
+// GetCountsPerIteration counts WIs including iteration-children and returns a map of iterationID->WICountsPerIteration
 func (r *GormWorkItemRepository) GetCountsPerIteration(ctx context.Context, spaceID uuid.UUID) (map[string]WICountsPerIteration, error) {
-	defer goa.MeasureSince([]string{"goa", "db", "workitem", "getCountPerIteration"}, time.Now())
+	defer goa.MeasureSince([]string{"goa", "db", "workitem", "getCountsPerIteration"}, time.Now())
+	db := r.db.Model(&iteration.Iteration{}).Where("space_id = ?", spaceID)
+	if db.Error != nil {
+		return nil, errors.NewInternalError(ctx, db.Error)
+	}
+	var allIterations []uuid.UUID
+	db.Pluck("id", &allIterations)
+	iterationTable := iteration.Iteration{}
+	iterationTableName := iterationTable.TableName()
+	iterationWithWICount := fmt.Sprintf(`
+	SELECT count(*) AS Total,
+		count(CASE fields->>'system.state'
+					WHEN 'closed' THEN '1'
+					ELSE NULL
+				END) AS Closed,
+		fields->>'system.iteration' AS iterationID
+	FROM %s wi
+	WHERE fields->>'system.iteration' IN
+		(SELECT id::text
+		FROM %s
+		WHERE space_id=?)
+	AND wi.deleted_at IS NULL
+	GROUP BY fields->>'system.iteration'`,
+		workitemTableName,
+		iterationTableName)
+	db = r.db.Raw(iterationWithWICount, spaceID.String())
 	var res []WICountsPerIteration
-	db := r.db.Table("work_items").Select(`iterations.id as IterationId, count(*) as Total,
-				count( case fields->>'system.state' when 'closed' then '1' else null end ) as Closed`).Joins(`left join iterations
-				on fields@> concat('{"system.iteration": "', iterations.id, '"}')::jsonb`).Where(`iterations.space_id = ?
-				and work_items.deleted_at IS NULL`, spaceID).Group(`IterationId`).Scan(&res)
+	db.Scan(&res)
 	if db.Error != nil {
+		log.Error(ctx, map[string]interface{}{
+			"space_id": spaceID.String(),
+			"err":      db.Error,
+		}, "unable to count WI for every iteration in a space")
 		return nil, errors.NewInternalError(ctx, db.Error)
 	}
+	wiMap := map[string]WICountsPerIteration{}
+	for _, r := range res {
+		wiMap[r.IterationID] = WICountsPerIteration{
+			IterationID: r.IterationID,
+			Total:       r.Total,
+			Closed:      r.Closed,
+		}
+	}
+	// put 0 count for iterations which are not in wiMap
+	// ToDo: Update count query to include non matching rows with 0 values
+	// Following operation can be skipped once above is done
+	for _, i := range allIterations {
+		if _, exists := wiMap[i.String()]; exists == false {
+			wiMap[i.String()] = WICountsPerIteration{
+				IterationID: i.String(),
+				Total:       0,
+				Closed:      0,
+			}
+		}
+	}
+
+	type IterationHavingChildrenID struct {
```

Comment:
I have not kept the type definition outside of the function that uses it, did you mean outside of the body?
done here https://github.com/almighty/almighty-core/pull/1151/files#diff-55f2da878ed2fcb42e5373010f02ff92R762